[E2E] Cluster upgrade from `8.8.0` to `8.9.0-SNAPSHOT` is broken

I don't think the problem is strictly related to ECK, but I'm opening an issue here as it is breaking several of our E2E tests.

When upgrading a cluster from `8.8.0` to `8.9.0-SNAPSHOT` (`imageID: docker.elastic.co/elasticsearch/elasticsearch@sha256:ab0645f7c629e6b730dc4eec9a3be3783cd18669f080742d2de0095066c058d5`) the first node that is restarted cannot join the cluster:

```json
{
	"@timestamp": "2023-06-04T23:20:43.280Z",
	"log.level": "INFO",
	"message": "failed to join {test-agent-upgrade-spn4-es-masterdata-0}{TuGMPMOjQB-yie2aiXLxpA}{PPGRWqpvTfGEE6x8qMPyZA}{test-agent-upgrade-spn4-es-masterdata-0}{10.92.161.51}{10.92.161.51:9300}{cdfhilmrstw}{8.8.0}{xpack.installed=true, ml.machine_memory=2147483648, ml.allocated_processors_double=8.0, k8s_node_name=gke-eck-e2e-stack-8-9-0--default-pool-02b08817-glh1, ml.allocated_processors=8, ml.max_jvm_size=1073741824} with JoinRequest{sourceNode={test-agent-upgrade-spn4-es-masterdata-2}{-KX0DBGvSriDROQNnvjaEA}{4wWMguynTUK-Jau9NEg8TA}{test-agent-upgrade-spn4-es-masterdata-2}{10.92.162.53}{10.92.162.53:9300}{cdfhilmrstw}{8.9.0}{xpack.installed=true, ml.machine_memory=2147483648, ml.allocated_processors_double=8.0, k8s_node_name=gke-eck-e2e-stack-8-9-0--default-pool-0470ebdd-3w03, ml.allocated_processors=8, ml.max_jvm_size=1073741824}, transportVersion=8500007, minimumTerm=1, optionalJoin=Optional.empty}",
	"ecs.version": "1.2.0",
	"service.name": "ES_ECS",
	"event.dataset": "elasticsearch.server",
	"process.thread.name": "elasticsearch[test-agent-upgrade-spn4-es-masterdata-2][transport_worker][T#2]",
	"log.logger": "org.elasticsearch.cluster.coordination.JoinHelper",
	"elasticsearch.node.name": "test-agent-upgrade-spn4-es-masterdata-2",
	"elasticsearch.cluster.name": "test-agent-upgrade-spn4",
	"error.type": "org.elasticsearch.transport.RemoteTransportException",
	"error.message": "[test-agent-upgrade-spn4-es-masterdata-0][10.92.161.51:9300][internal:cluster/coordination/join]",
	"error.stack_trace": "org.elasticsearch.transport.RemoteTransportException: [test-agent-upgrade-spn4-es-masterdata-0][10.92.161.51:9300][internal:cluster/coordination/join]\nCaused by: java.lang.IllegalStateException: failure when sending a join validation request from [{test-agent-upgrade-spn4-es-masterdata-0}{TuGMPMOjQB-yie2aiXLxpA}{PPGRWqpvTfGEE6x8qMPyZA}{test-agent-upgrade-spn4-es-masterdata-0}{10.92.161.51}{10.92.161.51:9300}{cdfhilmrstw}{8.8.0}] to [{test-agent-upgrade-spn4-es-masterdata-2}{-KX0DBGvSriDROQNnvjaEA}{4wWMguynTUK-Jau9NEg8TA}{test-agent-upgrade-spn4-es-masterdata-2}{10.92.162.53}{10.92.162.53:9300}{cdfhilmrstw}{8.9.0}]\n\tat org.elasticsearch.cluster.coordination.Coordinator.lambda$sendJoinValidate$13(Coordinator.java:730)\n\tat org.elasticsearch.action.ActionListenerImplementations$DelegatingResponseActionListener.acceptException(ActionListenerImplementations.java:126)\n\tat org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:59)\n\tat org.elasticsearch.action.ActionListenerImplementations$DelegatingResponseActionListener.onFailure(ActionListenerImplementations.java:131)\n\tat org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:59)\n\tat org.elasticsearch.action.ActionListenerImplementations.safeOnFailure(ActionListenerImplementations.java:71)\n\tat org.elasticsearch.action.DelegatingActionListener.onFailure(DelegatingActionListener.java:27)\n\tat org.elasticsearch.action.ActionListenerImplementations$RunAfterActionListener.onFailure(ActionListenerImplementations.java:181)\n\tat org.elasticsearch.action.ActionListenerResponseHandler.handleException(ActionListenerResponseHandler.java:48)\n\tat org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleException(TransportService.java:1405)\n\tat org.elasticsearch.transport.InboundHandler.doHandleException(InboundHandler.java:461)\n\tat org.elasticsearch.transport.InboundHandler$3.doRun(InboundHandler.java:453)\n\tat org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:983)\n\tat org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)\n\tat java.lang.Thread.run(Thread.java:1623)\nCaused by: org.elasticsearch.transport.RemoteTransportException: [test-agent-upgrade-spn4-es-masterdata-2][10.92.162.53:9300][internal:cluster/coordination/join/validate]\nCaused by: java.lang.IllegalStateException: unexpected byte [0x1d]\n\tat org.elasticsearch.common.io.stream.StreamInput.readBoolean(StreamInput.java:564)\n\tat org.elasticsearch.common.io.stream.StreamInput.readBoolean(StreamInput.java:554)\n\tat org.elasticsearch.common.io.stream.StreamInput.readOptionalWriteable(StreamInput.java:1005)\n\tat org.elasticsearch.cluster.metadata.DataStream.<init>(DataStream.java:763)\n\tat org.elasticsearch.common.io.stream.StreamInput.readImmutableOpenMap(StreamInput.java:737)\n\tat org.elasticsearch.cluster.metadata.DataStreamMetadata.<init>(DataStreamMetadata.java:91)\n\tat org.elasticsearch.common.io.stream.NamedWriteableAwareStreamInput.readNamedWriteable(NamedWriteableAwareStreamInput.java:56)\n\tat org.elasticsearch.common.io.stream.NamedWriteableAwareStreamInput.readNamedWriteable(NamedWriteableAwareStreamInput.java:32)\n\tat org.elasticsearch.cluster.metadata.Metadata.readFrom(Metadata.java:1653)\n\tat org.elasticsearch.cluster.ClusterState.readFrom(ClusterState.java:907)\n\tat org.elasticsearch.cluster.coordination.ValidateJoinRequest.readCompressed(ValidateJoinRequest.java:59)\n\tat org.elasticsearch.cluster.coordination.ValidateJoinRequest.lambda$new$0(ValidateJoinRequest.java:36)\n\tat org.elasticsearch.cluster.coordination.ValidateJoinRequest.getOrReadState(ValidateJoinRequest.java:76)\n\tat org.elasticsearch.cluster.coordination.JoinValidationService.lambda$new$2(JoinValidationService.java:121)\n\tat org.elasticsearch.xpack.security.transport.SecurityServerTransportInterceptor$ProfileSecuredRequestHandler$1.doRun(SecurityServerTransportInterceptor.java:570)\n\tat org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)\n\tat org.elasticsearch.xpack.security.transport.SecurityServerTransportInterceptor$ProfileSecuredRequestHandler$3.onResponse(SecurityServerTransportInterceptor.java:632)\n\tat org.elasticsearch.xpack.security.transport.SecurityServerTransportInterceptor$ProfileSecuredRequestHandler$3.onResponse(SecurityServerTransportInterceptor.java:621)\n\tat org.elasticsearch.xpack.security.authz.AuthorizationService.authorizeSystemUser(AuthorizationService.java:655)\n\tat org.elasticsearch.xpack.security.authz.AuthorizationService.authorize(AuthorizationService.java:309)\n\tat org.elasticsearch.xpack.security.transport.ServerTransportFilter.lambda$inbound$1(ServerTransportFilter.java:113)\n\tat org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:158)\n\tat org.elasticsearch.action.ActionListenerImplementations$MappedActionListener.onResponse(ActionListenerImplementations.java:93)\n\tat org.elasticsearch.xpack.security.authc.AuthenticatorChain.authenticateAsync(AuthenticatorChain.java:94)\n\tat org.elasticsearch.xpack.security.authc.AuthenticationService.authenticate(AuthenticationService.java:261)\n\tat org.elasticsearch.xpack.security.authc.AuthenticationService.authenticate(AuthenticationService.java:199)\n\tat org.elasticsearch.xpack.security.transport.ServerTransportFilter.authenticate(ServerTransportFilter.java:126)\n\tat org.elasticsearch.xpack.security.transport.ServerTransportFilter.inbound(ServerTransportFilter.java:104)\n\tat org.elasticsearch.xpack.security.transport.SecurityServerTransportInterceptor$ProfileSecuredRequestHandler.messageReceived(SecurityServerTransportInterceptor.java:643)\n\tat org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:74)\n\tat org.elasticsearch.transport.InboundHandler$1.doRun(InboundHandler.java:315)\n\tat org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:983)\n\tat org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)\n\tat java.lang.Thread.run(Thread.java:1623)\n"
}
```

Logs on `test-agent-upgrade-spn4-es-masterdata-0`:

```json
{
	"@timestamp": "2023-06-04T23:24:55.366Z",
	"log.level": "WARN",
	"message": "failed to validate incoming join request from node [{test-agent-upgrade-spn4-es-masterdata-2}{-KX0DBGvSriDROQNnvjaEA}{4wWMguynTUK-Jau9NEg8TA}{test-agent-upgrade-spn4-es-masterdata-2}{10.92.162.53}{10.92.162.53:9300}{cdfhilmrstw}{8.9.0}{k8s_node_name=gke-eck-e2e-stack-8-9-0--default-pool-0470ebdd-3w03, ml.machine_memory=2147483648, xpack.installed=true, ml.allocated_processors_double=8.0, ml.allocated_processors=8, ml.max_jvm_size=1073741824}]",
	"ecs.version": "1.2.0",
	"service.name": "ES_ECS",
	"event.dataset": "elasticsearch.server",
	"process.thread.name": "elasticsearch[test-agent-upgrade-spn4-es-masterdata-0][cluster_coordination][T#1]",
	"log.logger": "org.elasticsearch.cluster.coordination.Coordinator",
	"elasticsearch.cluster.uuid": "kggbEqh_RiOTtKgcB4d3sQ",
	"elasticsearch.node.id": "TuGMPMOjQB-yie2aiXLxpA",
	"elasticsearch.node.name": "test-agent-upgrade-spn4-es-masterdata-0",
	"elasticsearch.cluster.name": "test-agent-upgrade-spn4",
	"error.type": "org.elasticsearch.transport.RemoteTransportException",
	"error.message": "[test-agent-upgrade-spn4-es-masterdata-2][10.92.162.53:9300][internal:cluster/coordination/join/validate]",
	"error.stack_trace": "org.elasticsearch.transport.RemoteTransportException: [test-agent-upgrade-spn4-es-masterdata-2][10.92.162.53:9300][internal:cluster/coordination/join/validate]\nCaused by: java.lang.IllegalStateException: unexpected byte [0x1d]\n\tat org.elasticsearch.common.io.stream.StreamInput.readBoolean(StreamInput.java:564)\n\tat org.elasticsearch.common.io.stream.StreamInput.readBoolean(StreamInput.java:554)\n\tat org.elasticsearch.common.io.stream.StreamInput.readOptionalWriteable(StreamInput.java:1005)\n\tat org.elasticsearch.cluster.metadata.DataStream.<init>(DataStream.java:763)\n\tat org.elasticsearch.common.io.stream.StreamInput.readImmutableOpenMap(StreamInput.java:737)\n\tat org.elasticsearch.cluster.metadata.DataStreamMetadata.<init>(DataStreamMetadata.java:91)\n\tat org.elasticsearch.common.io.stream.NamedWriteableAwareStreamInput.readNamedWriteable(NamedWriteableAwareStreamInput.java:56)\n\tat org.elasticsearch.common.io.stream.NamedWriteableAwareStreamInput.readNamedWriteable(NamedWriteableAwareStreamInput.java:32)\n\tat org.elasticsearch.cluster.metadata.Metadata.readFrom(Metadata.java:1653)\n\tat org.elasticsearch.cluster.ClusterState.readFrom(ClusterState.java:907)\n\tat org.elasticsearch.cluster.coordination.ValidateJoinRequest.readCompressed(ValidateJoinRequest.java:59)\n\tat org.elasticsearch.cluster.coordination.ValidateJoinRequest.lambda$new$0(ValidateJoinRequest.java:36)\n\tat org.elasticsearch.cluster.coordination.ValidateJoinRequest.getOrReadState(ValidateJoinRequest.java:76)\n\tat org.elasticsearch.cluster.coordination.JoinValidationService.lambda$new$2(JoinValidationService.java:121)\n\tat org.elasticsearch.xpack.security.transport.SecurityServerTransportInterceptor$ProfileSecuredRequestHandler$1.doRun(SecurityServerTransportInterceptor.java:570)\n\tat org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)\n\tat org.elasticsearch.xpack.security.transport.SecurityServerTransportInterceptor$ProfileSecuredRequestHandler$3.onResponse(SecurityServerTransportInterceptor.java:632)\n\tat org.elasticsearch.xpack.security.transport.SecurityServerTransportInterceptor$ProfileSecuredRequestHandler$3.onResponse(SecurityServerTransportInterceptor.java:621)\n\tat org.elasticsearch.xpack.security.authz.AuthorizationService.authorizeSystemUser(AuthorizationService.java:655)\n\tat org.elasticsearch.xpack.security.authz.AuthorizationService.authorize(AuthorizationService.java:309)\n\tat org.elasticsearch.xpack.security.transport.ServerTransportFilter.lambda$inbound$1(ServerTransportFilter.java:113)\n\tat org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:158)\n\tat org.elasticsearch.action.ActionListenerImplementations$MappedActionListener.onResponse(ActionListenerImplementations.java:93)\n\tat org.elasticsearch.xpack.security.authc.AuthenticatorChain.authenticateAsync(AuthenticatorChain.java:94)\n\tat org.elasticsearch.xpack.security.authc.AuthenticationService.authenticate(AuthenticationService.java:261)\n\tat org.elasticsearch.xpack.security.authc.AuthenticationService.authenticate(AuthenticationService.java:199)\n\tat org.elasticsearch.xpack.security.transport.ServerTransportFilter.authenticate(ServerTransportFilter.java:126)\n\tat org.elasticsearch.xpack.security.transport.ServerTransportFilter.inbound(ServerTransportFilter.java:104)\n\tat org.elasticsearch.xpack.security.transport.SecurityServerTransportInterceptor$ProfileSecuredRequestHandler.messageReceived(SecurityServerTransportInterceptor.java:643)\n\tat org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:74)\n\tat org.elasticsearch.transport.InboundHandler$1.doRun(InboundHandler.java:315)\n\tat org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:983)\n\tat org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)\n\tat java.lang.Thread.run(Thread.java:1623)\n"
}
```

--- a/test/e2e/test/version.go
+++ b/test/e2e/test/version.go
@@ -31,21 +31,6 @@ func SkipInvalidUpgrade(t *testing.T, srcVersion string, dstVersion string) {
 	if !isValid {
 		t.SkipNow()
 	}
-
-	// Temporarily skip upgrades to `8.9.x-SNAPSHOT`
-	dstVer := version.MustParse(dstVersion)
-	if dstVer.Major == 8 && dstVer.Minor == 9 && isSnapshot(dstVer) {
-		t.Skip("Upgrades to 8.9.0-SNAPSHOT are temporarily disabled, refer to https://github.com/elastic/cloud-on-k8s/issues/6878")
-	}
-}
-
-func isSnapshot(v version.Version) bool {
-	for _, pre := range v.Pre {
-		if pre.VersionStr == "SNAPSHOT" {
-			return true
-		}
-	}
-	return false
 }
 
 // isValidUpgrade reports whether an upgrade from one version to another version is valid.

